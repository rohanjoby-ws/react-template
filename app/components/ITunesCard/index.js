/**
 *
 * ITunesCard
 *
 */

import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import truncate from 'lodash/truncate';
import { isEmpty } from 'lodash';
import { Card, Popover, Button, Progress } from 'antd';
import { PlayCircleOutlined, PauseCircleOutlined } from '@ant-design/icons';
import T from '@components/T';
import If from '@components/If';
import { colors } from '@app/themes';
import { ACTIONS } from '@utils/constants';

const CustomCard = styled(Card)`
  && {
    margin: 1.25em 0;
    box-shadow: ${colors.boxShadow};
  }
`;
const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 0.75rem;
  gap: 1em;
`;
const CustomImg = styled.img`
  display: block;
  margin: 0.75rem auto;
  height: 10rem;
`;

export function ITunesCard({ track, onActionClick }) {
  const audioRef = useRef(null);
  const [progress, setProgress] = useState(0);
  const [intervalStore, setIntervalStore] = useState(null);

  useEffect(() => {
    return () => {
      clearInterval(intervalStore);
    };
  }, []);
  useEffect(() => {
    clearInterval(intervalStore);
  }, [track]);

  function calculateProgress() {
    const intervalValue = setInterval(() => {
      const fraction = (audioRef.current.currentTime / audioRef.current.duration) * 100;
      setProgress(fraction);
    }, 1000);
    setIntervalStore(intervalValue);
  }
  const handleTrackPlay = (action) => {
    if (action === ACTIONS.PLAY) {
      calculateProgress();
    } else {
      clearInterval(intervalStore);
    }
    onActionClick(audioRef, action);
  };

  const text = <T id="details" type="subheading" marginBottom={5} />;
  const content = (
    <>
      <If condition={track?.trackPrice} otherwise={<T data-testid="price-unavailable" id="track_price_unavailable" />}>
        <T id="track-price" values={{ price: track?.trackPrice }} />
      </If>
      <If
        condition={!isEmpty(track?.primaryGenreName)}
        otherwise={<T data-testid="genre-unavailable" id="track_genre_unavailable" />}
      >
        <T id="genre-name" values={{ name: track?.primaryGenreName }} />
      </If>
    </>
  );
  return (
    <CustomCard data-testid="itunes-card">
      <If
        condition={!isEmpty(track?.collectionName)}
        otherwise={<T data-testid="collection-unavailable" id="collection_name_unavailable" />}
      >
        <T
          data-testid="collectionName"
          id="collection-name"
          values={{ name: truncate(track?.collectionName, { length: 28 }) }}
        />
      </If>
      <Popover placement="topRight" title={text} content={content}>
        <CustomImg src={track.artworkUrl100} alt="artwork" />
      </Popover>
      <Wrapper>
        <Button data-testid="play-button" onClick={() => handleTrackPlay(ACTIONS.PLAY)} type="text">
          <PlayCircleOutlined style={{ fontSize: '20px' }} />
        </Button>
        <Progress percent={progress} showInfo={false} data-testid="progress-bar" />
        <Button data-testid="pause-button" onClick={() => handleTrackPlay(ACTIONS.PAUSE)} type="text">
          <PauseCircleOutlined style={{ fontSize: '20px' }} />
        </Button>
      </Wrapper>
      <audio data-testid="audio-element" ref={audioRef} src={track.previewUrl} onClick={handleTrackPlay} />
      <If
        condition={!isEmpty(track?.trackName)}
        otherwise={<T data-testid="track-unavailable" id="track_name_unavailable" />}
      >
        <T
          data-testid="trackName"
          id="track-name"
          values={{ name: truncate(track?.trackName, { length: 28 }) }}
          type="subheading"
          marginBottom={5}
        />
      </If>
      <If
        condition={!isEmpty(track?.artistName)}
        otherwise={<T data-testid="artist-unavailable" id="artist_unavailable" />}
      >
        <T data-testid="artistName" id="artist-name" values={{ name: truncate(track?.artistName, { length: 35 }) }} />
      </If>
    </CustomCard>
  );
}

ITunesCard.propTypes = {
  track: PropTypes.object.isRequired,
  onActionClick: PropTypes.func
};

export default ITunesCard;
