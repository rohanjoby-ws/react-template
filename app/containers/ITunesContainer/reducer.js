/*
 *
 * ITunesContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const initialState = {
  searchQuery: null,
  iTunesData: {},
  iTunesError: null,

  trackID: null,
  trackData: {},
  trackError: null
};

export const { Types: iTunesContainerTypes, Creators: iTunesContainerCreators } = createActions({
  requestGetITunesSongs: ['searchQuery'],
  successGetITunesSongs: ['data'],
  failureGetITunesSongs: ['error'],
  clearGetITunesSongs: {},

  //requestGetTrackData: ['trackID'],
  successGetTrackData: ['data'],
  failureGetTrackData: ['error'],
  clearGetTrackData: {}
});

export const iTunesContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case iTunesContainerTypes.REQUEST_GET_I_TUNES_SONGS:
        draft.searchQuery = action.searchQuery;
        break;
      case iTunesContainerTypes.CLEAR_GET_I_TUNES_SONGS:
        draft.searchQuery = null;
        draft.iTunesError = null;
        draft.iTunesData = {};
        break;
      case iTunesContainerTypes.SUCCESS_GET_I_TUNES_SONGS:
        draft.iTunesData = action.data;
        break;
      case iTunesContainerTypes.FAILURE_GET_I_TUNES_SONGS:
        draft.iTunesError = get(action.error, 'message', 'something_went_wrong');
        break;

      // case iTunesContainerTypes.REQUEST_GET_TRACK_DATA:
      //   draft.trackId = action.trackID;
      //   break;
      case iTunesContainerTypes.CLEAR_GET_TRACK_DATA:
        draft.trackId = null;
        draft.trackError = null;
        draft.trackData = {};
        break;
      case iTunesContainerTypes.SUCCESS_GET_TRACK_DATA:
        draft.trackData = action.data;
        break;
      case iTunesContainerTypes.FAILURE_GET_TRACK_DATA:
        draft.trackError = get(action.error, 'message', 'something_went_wrong');
        break;

      default:
        return state;
    }
  });
