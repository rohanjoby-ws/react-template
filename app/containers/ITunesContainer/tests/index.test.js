/**
 *
 * Tests for ITunesContainer
 *
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { timeout, renderProvider } from '@utils/testUtils';
import { translate } from '@app/components/IntlGlobalProvider';
import { ITunesContainerTest as ITunesContainer, mapDispatchToProps } from '../index';
import { iTunesContainerTypes } from '../reducer';

describe('<ITunesContainer /> container tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<ITunesContainer />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchClearITunesSongs on empty change', async () => {
    const getITunesSongsSpy = jest.fn();
    const clearITunesSongsSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <ITunesContainer dispatchClearITunesSongs={clearITunesSongsSpy} dispatchITunesSongs={getITunesSongsSpy} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: 'a' }
    });
    await timeout(500);
    expect(getITunesSongsSpy).toBeCalled();
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: '' }
    });
    await timeout(500);
    expect(clearITunesSongsSpy).toBeCalled();
  });

  it('should call dispatchITunesSongs on change and after enter', async () => {
    const searchQuery = 'Infinity';
    const { getByTestId } = renderProvider(<ITunesContainer dispatchITunesSongs={submitSpy} />);
    const searchBar = getByTestId('search-bar');
    fireEvent.change(searchBar, {
      target: { value: searchQuery }
    });
    await timeout(500);
    expect(submitSpy).toBeCalledWith(searchQuery);

    fireEvent.keyDown(searchBar, {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
    expect(submitSpy).toBeCalledWith(searchQuery);
  });

  it('should call dispatchITunesSongs on submit', async () => {
    const searchQuery = 'Infinity';
    const { getByTestId } = renderProvider(<ITunesContainer dispatchITunesSongs={submitSpy} />);
    fireEvent.keyDown(getByTestId('search-bar'), { keyCode: 13, target: { value: searchQuery } });

    await timeout(500);
    expect(submitSpy).toBeCalledWith(searchQuery);
  });

  it('should  dispatchiTunesSongs on update on mount if searchQuery is already persisted', async () => {
    const searchQuery = 'Infinity';
    renderProvider(<ITunesContainer searchQuery={searchQuery} iTunesData={null} dispatchITunesSongs={submitSpy} />);

    await timeout(500);
    expect(submitSpy).toBeCalledWith(searchQuery);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchITunesSongsSpy = jest.fn();
    const searchQuery = 'Infinity';
    const actions = {
      dispatchITunesSongs: { searchQuery, type: iTunesContainerTypes.REQUEST_GET_I_TUNES_SONGS },
      dispatchClearITunesSongs: { type: iTunesContainerTypes.CLEAR_GET_I_TUNES_SONGS }
    };

    const props = mapDispatchToProps(dispatchITunesSongsSpy);
    props.dispatchITunesSongs(searchQuery);
    expect(dispatchITunesSongsSpy).toHaveBeenCalledWith(actions.dispatchITunesSongs);

    await timeout(500);
    props.dispatchClearITunesSongs();
    expect(dispatchITunesSongsSpy).toHaveBeenCalledWith(actions.dispatchClearITunesSongs);
  });

  it('should render default error message when search goes wrong', () => {
    const defaultError = translate('something_went_wrong');
    const { getByTestId } = renderProvider(<ITunesContainer iTunesError={defaultError} />);
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });

  it('should render the data when loading becomes false', () => {
    const iTunesData = { totalCount: 1, results: [{ artistName: 'Jaymes Young' }] };
    const { getByTestId } = renderProvider(<ITunesContainer iTunesData={iTunesData} dispatchITunesSongs={submitSpy} />);
    expect(getByTestId('for')).toBeInTheDocument();
  });

  it('should render exact number of iTuneCards as per totalCount in result', () => {
    const totalCount = 2;
    const iTunesData = {
      totalCount,
      results: [
        {
          artistName: 'Mark Ronson',
          trackName: 'Uptown Funk (feat. Bruno Mars)',
          primaryGenreName: 'Pop'
        },
        {
          artistName: 'Dynamix Music',
          trackName: 'Uptown Funk',
          primaryGenreName: 'Fitness & Workout'
        }
      ]
    };
    const { getAllByTestId } = renderProvider(
      <ITunesContainer iTunesData={iTunesData} dispatchITunesSongs={submitSpy} />
    );
    expect(getAllByTestId('itunes-card').length).toBe(totalCount);
  });

  it('should render Skeleton Comp when "loading" is true', async () => {
    const searchQuery = 'Infinity';
    const { getByTestId, baseElement } = renderProvider(
      <ITunesContainer dispatchITunesSongs={submitSpy} searchQuery={searchQuery} />
    );
    fireEvent.change(getByTestId('search-bar'), { target: { value: searchQuery } });
    await timeout(500);
    expect(baseElement.getElementsByClassName('ant-skeleton').length).toBe(1);
  });
});
