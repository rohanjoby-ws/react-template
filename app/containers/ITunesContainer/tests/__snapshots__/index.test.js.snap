// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ITunesContainer /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      ITunesContainer text
    </div>
  </div>
</body>
`;
